lib/matplotlib/backends/qt_compat.py

```diff
@@ -1,220 +1,159 @@
-            QT_API = QT_API_PYQT
-            QT_RC_MAJOR_VERSION = 4
-        else:
-            raise
-
-if _sip_imported:
-    # needs to be if so we can re-test the value of QT_API which may
-    # have been changed in the above if block
-    if QT_API in [QT_API_PYQT, QT_API_PYQTv2]:  # PyQt4 API
-        from PyQt4 import QtCore, QtGui
-
-        try:
-            if sip.getapi("QString") > 1:
-                # Use new getSaveFileNameAndFilter()
-                _getSaveFileName = QtGui.QFileDialog.getSaveFileNameAndFilter
-            else:
-
-                # Use old getSaveFileName()
-                def _getSaveFileName(*args, **kwargs):
-                    return (QtGui.QFileDialog.getSaveFileName(*args, **kwargs),
-                            None)
-
-        except (AttributeError, KeyError):
-
-            # call to getapi() can fail in older versions of sip
-            def _getSaveFileName(*args, **kwargs):
-                return QtGui.QFileDialog.getSaveFileName(*args, **kwargs), None
-
-
-if QT_API == QT_API_PYSIDE2:
-    try:
+        __version__ = QtCore.PYQT_VERSION_STR
+        QtCore.Signal = QtCore.pyqtSignal
+        QtCore.Slot = QtCore.pyqtSlot
+        QtCore.Property = QtCore.pyqtProperty
+    elif QT_API == QT_API_PYSIDE2:
         from PySide2 import QtCore, QtGui, QtWidgets, __version__
-        _getSaveFileName = QtWidgets.QFileDialog.getSaveFileName
-    except ImportError:
-        # tried PySide2, failed, fall back to PySide
-        QT_RC_MAJOR_VERSION = 4
-        QT_API = QT_API_PYSIDE
-
-if QT_API == QT_API_PYSIDE:  # try importing pyside
-    try:
-        from PySide import QtCore, QtGui, __version__, __version_info__
-    except ImportError:
-        raise ImportError(
-            "Matplotlib qt-based backends require an external PyQt4, PyQt5,\n"
-            "PySide or PySide2 package to be installed, but it was not found.")
+    else:
+        raise ValueError("Unexpected value for the 'backend.qt5' rcparam")
+    _getSaveFileName = QtWidgets.QFileDialog.getSaveFileName
 
-    if __version_info__ < (1, 0, 3):
-        raise ImportError(
-            "Matplotlib backend_qt4 and backend_qt4agg require PySide >=1.0.3")
+    def is_pyqt5():
+        return True
 
-    _getSaveFileName = QtGui.QFileDialog.getSaveFileName
 
+def _setup_pyqt4():
+    global QtCore, QtGui, QtWidgets, __version__, is_pyqt5, _getSaveFileName
 
-if QT_API in (QT_API_PYQT, QT_API_PYQTv2, QT_API_PYQT5):
-    # Alias PyQt-specific functions for PySide compatibility.
-    QtCore.Signal = QtCore.pyqtSignal
-    try:
+    def _setup_pyqt4_internal(api):
+        global QtCore, QtGui, QtWidgets, \
+            __version__, is_pyqt5, _getSaveFileName
+        # List of incompatible APIs:
+        # http://pyqt.sourceforge.net/Docs/PyQt4/incompatible_apis.html
+        _sip_apis = ["QDate", "QDateTime", "QString", "QTextStream", "QTime",
+                     "QUrl", "QVariant"]
+        import sip
```

Comment:
move `import sip` into the try-except as done in @tacaswell s commit to #11600.

Reply:
Not clear why this would be needed?  The way this is currently written, if QT_API is set then we try that binding exactly (and fail with ImportError if not available, which includes sip not being available when PyQt4 is requested); if not (QT_API is None, which would only occur in a fully manual embedding while rcParams["backend"] is not even "qt4agg" or "qt5agg"), then we just try everything in the loop at the bottom where each iteration is protected by a try... except ImportError.

Reply:
Because riverbank (in the _last_ planned pyqt4 release) made sip a 'private' library such that before you import pyqt `import sip` fails.

I am a bit unclear how to set the sip api in that case, but...

Reply:
I thought that was pyqt5?

Reply:
I am not sure what you refer to; I can't repro with PyQt4 installed from conda (it's not available as a pypi wheel) or PyQt5 from either conda or pypi.